### PR TITLE
Add querynode plugin

### PR DESCRIPTION
**Pull request by chasingegg**
issue: #22372

### CODE REVIEW

File: internal/querynode/query_node.go [hunk 1]
```diff
@@ -135,6 +138,42 @@ func GetQueryNode() *QueryNode {
...
+func initHook() Hook {
+	path := Params.QueryNodeCfg.SoPath.GetValue()
+	if path == "" {
+		return nil
+	}
+	log.Debug("start to load plugin", zap.String("path", path))
+	p, err := plugin.Open(path)
+	if err != nil {
+		log.Error("fail to open the plugin, error: ", zap.Error(err))
+		return nil
```
Comment (congqixia) [inline internal/querynode/query_node.go:155]: if load plugin failed, is it ok to ignore here?

File: internal/querynode/query_node.go [hunk 2]
```diff
@@ -135,6 +138,42 @@ func GetQueryNode() *QueryNode {
...
+	}
+	log.Debug("plugin open")
+
+	h, err := p.Lookup("QueryNodePlugin")
+	if err != nil {
+		log.Error("fail to the 'QueryNodePlugin' object in the plugin, error: ", zap.Error(err))
+		return nil
```
Comment (congqixia) [inline internal/querynode/query_node.go:162]: Ditto

File: internal/querynode/impl.go [hunk 3]
```diff
@@ -887,6 +889,40 @@ func (node *QueryNode) searchWithDmlChannel(ctx context.Context, req *querypb.Se
 	)
 	defer cancel()
 
+	// optimize search params
+	if req.Req.GetSerializedExprPlan() != nil && node.queryHook != nil {
+		segmentNum := len(cluster.segments) * channelNum
```
Comment (congqixia) [inline internal/querynode/impl.go:894]: `channelNum` is not accurate if it means the total channel number of one collection
also, this line accesses `cluster.segments` without mutex protection

Comment (chasingegg) [reply internal/querynode/impl.go:894]: channelNum is got from len(req.GetDmlChannels()), how we get the  correct total channel number of one collection?

Comment (congqixia) [reply internal/querynode/impl.go:894]: `req.GetDmlChannels()` returns channel number in this request. If channels are loaded in different querynode, this length slice shall be 1. For now, shard leader does not contains information for shard number of one collection.

**Comment by chasingegg** [comment]
rerun ut

**Comment by chasingegg** [comment]
rerun ut

**Comment by chasingegg** [comment]
rerun ut

**Comment by chasingegg** [comment]
rerun ut

File: internal/querynode/query_node.go (hunk 1 above)
Comment (chasingegg) [reply internal/querynode/query_node.go:155]: yeah, still use the original search parameter provided by user

File: internal/querynode/impl.go (hunk 3 above)
Comment (chasingegg) [reply internal/querynode/impl.go]: add totalchannelnum to SearchRequest proto, help review again

File: internal/querynode/query_node.go (hunk 1 above)
Comment (chasingegg) [reply internal/querynode/query_node.go]: if autoindex enabled, here we need to panic directly

**Comment by chasingegg** [comment]
rerun ut

**Comment by chasingegg** [comment]
cc @congqixia

File: internal/util/indexparams/disk_index_params.go [hunk 4]
```diff
@@ -52,6 +53,97 @@ func getRowDataSizeOfFloatVector(numRows int64, dim int64) int64 {
...
+}
+
+const (
+	BuildRatioKey                   = "build_ratio"
+	PrepareRatioKey                 = "prepare_ratio"
+	DefaultPQCodeBudgetGBRatio      = 0.125
+	DefaultBuildNumThreadsRatio     = 1.0
+	DefaultSearchCacheBudgetGBRatio = 0.10
+	DefaultLoadNumThreadRatio       = 8.0
+	DefaultBeamWidthRatio           = 4.0
```
Comment (congqixia) [inline internal/util/indexparams/disk_index_params.go:71]: NIT, TODO: Configurable

File: internal/querynode/query_node.go [hunk 5]
```diff
@@ -147,6 +183,26 @@ func NewQueryNode(ctx context.Context, factory dependency.Factory) *QueryNode {
 		lifetime:            lifetime.NewLifetime(commonpb.StateCode_Abnormal),
 	}
 
+	// init querynode plugin
+	var err error
+	queryNode.queryHook, err = initHook()
+	if err != nil {
+		log.Error(err.Error())
```
Comment (congqixia) [inline internal/querynode/query_node.go:190]: NIT, TODO: `log.Error(msg, zap.Error(err))`

File: internal/querynode/query_node.go [hunk 6]
```diff
@@ -135,6 +138,39 @@ func GetQueryNode() *QueryNode {
...
+	}
+	log.Debug("start to load plugin", zap.String("path", path))
+
+	p, err := plugin.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("fail to open the plugin, error: %s", err.Error())
+	}
+	log.Debug("plugin open")
+
+	h, err := p.Lookup("QueryNodePlugin")
```
Comment (congqixia) [inline internal/querynode/query_node.go:159]: NIT, TODO: use common const

File: internal/querynode/impl.go [hunk 7]
```diff
@@ -910,6 +912,44 @@ func (node *QueryNode) searchWithDmlChannel(ctx context.Context, req *querypb.Se
...
+				"topk":         qinfo.GetTopk(),
+				"search_param": qinfo.GetSearchParams(),
+				"segment_num":  segmentNum,
+			}
+			err := node.queryHook.Run(paramMap)
+			if err != nil {
+				failRet.Status.Reason = err.Error()
+				return failRet, nil
+			}
+			qinfo.Topk = paramMap["topk"].(int64)
```
Comment (congqixia) [inline internal/querynode/impl.go:941]: NIT, TODO: use common const key

**Review by congqixia — APPROVED** [review]
Please fix the comments later. Overall lgtm
/lgtm

**Pull request merged**